xcdat/averager.py

```diff
@@ -0,0 +1,458 @@
+#         _averager(): perform weighted average (with appropriate masking)
+
+# %% Notes
+
+# Differences between CDAT and this algorithm are smaller than 10**-4
+# CDAT is faster for smaller spatial scales, but this algorithm is faster for larger spatial scales
+# Using xarrays data reduction functionality gives similar results, but is slower than the averager function defined here
+# Averaging axes are specified with a list (e.g., axis=['lat', 'lon']) rather than the CDAT convention (axis='xy')
+# It would be helpful to be able to get axis weights (get_axis_weights()) more generically (some information is hardcoded)
+    # If .bounds had a method to return a list of bounds entries, we could create weights for all dimensions
+# May want to keep the axis weights or full weight array (i.e., add this information into the variable dataarray properties)
+# Is xvariable a good name?
+# Need to write unit tests
+# Can remove logging around bounds found in dataset
+# Functions are too specific to "lat" and "lon" (may have issues if dims are latitude / longitude)
+# Should remove these notes and outline...
+
+
+def _validate_region_lon_bounds(grid_lon_bounds, region_lon_bounds):
+    """Determines if values in a list of longitude coordinates falls within
+    longitudinal grid cell boundaries of a specified longitude axis.
+
+    Operations include:
+
+    - Check to determine whether longitude values are within the boundaries of
+    the target longitude axis
+
+    Parameters
+    ----------
+    grid_lon_bounds : xr.DataArray
+        longitude axis [len(lon), 2]
+    region_lon_bounds : float | int | list | array
+        list of longitude values to validate
+
+    Returns
+    -------
+    boolean
+        True if all input longitude bounds are within the boundaries of the
+        target longitudinal axis
+
+    Examples
+    --------
+    Import:
+    >>> from xcdat.dataset import open_dataset
+    >>> from xcdat.variable import open_variable
+
+    Open dataset and dataarray:
+    >>> filename = 'test.nc'
+    >>> ds = open_dataset(filename)
+    >>> tas = open_variable(ds, 'tas')
+
+    Specify some regional bounds:
+    >>> regional_bounds = [-170, -120]
+
+    Validate:
+    >>> in_bounds = _validate_region_lon_bounds(tas.bounds.lon,
+                                                regional_bounds)
+    """
+    # ensure iterable array
+    region_lon_bounds = np.asarray(region_lon_bounds)
+    region_lon_bounds = np.atleast_1d(region_lon_bounds)
+    # get the max / min bounds for each grid cell
+    min_grid_lon_bounds = np.min(grid_lon_bounds, axis=1)
+    max_grid_lon_bounds = np.max(grid_lon_bounds, axis=1)
+    # iterate over regional bounds and check to see if they fall within
+    # any grid cell bounds.
+    in_bounds = True
+    for lon in region_lon_bounds:
+        inds = np.where((min_grid_lon_bounds <= lon) &
+                        (max_grid_lon_bounds >= lon))[0]
+        if len(inds) == 0:
+            in_bounds = False
```

Comment:
This may be a performance bottleneck since it is looping through each element of the ndarray, then performing a call against two numpy ndarrays using a `.where()`.